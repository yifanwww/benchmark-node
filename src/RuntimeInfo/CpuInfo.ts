import os from 'os';

import { exec } from '../Tools/exec';

export class CpuInfo {
    public name: string;
    public processors: number;
    public logicalCores: number;
    public plysicalCores: number;

    public constructor() {
        this.name = os.cpus()[0].model.trim();

        this.processors = this.getNumberOfProcesors();

        this.logicalCores = os.cpus().length;
        this.plysicalCores = this.getPlysicalCores();
    }

    private getNumberOfProcesors(): number {
        switch (process.platform) {
            case 'linux': {
                const output = exec('lscpu | egrep Socket');
                const index = output.indexOf(':');
                return Number.parseInt(output.slice(index + 1).trim(), 10);
            }

            case 'darwin': {
                const output = exec('system_profiler SPHardwareDataType | grep Processors');
                const index = output.indexOf(':');
                return Number.parseInt(output.slice(index + 1).trim(), 10);
            }

            case 'win32': {
                const output = exec('WMIC CPU Get Name');
                return output.split(os.EOL).length - 1;
            }

            default:
                return 0;
        }
    }

    private getPlysicalCores(): number {
        switch (process.platform) {
            case 'linux': {
                const output = exec('lscpu -p | egrep -v "^#" | sort -u -t, -k 2,4 | wc -l');
                return Number.parseInt(output, 10);
            }

            case 'darwin': {
                const output = exec('sysctl -n hw.physicalcpu_max');
                return Number.parseInt(output, 10);
            }

            case 'win32': {
                const output = exec('WMIC CPU Get NumberOfCores');
                return output
                    .split(os.EOL)
                    .slice(1)
                    .map((line) => Number.parseInt(line.trim(), 10))
                    .reduce((sum, number) => sum + number, 0);
            }

            default:
                return 0;
        }
    }

    public toString(): string {
        const processorsDesc = this.processors > 0 ? `${this.processors} CPU` : '';

        let coresDesc;
        if (this.logicalCores === 0 && this.plysicalCores === 0) {
            coresDesc = '';
        } else if (this.logicalCores === 1 && this.plysicalCores === 0) {
            coresDesc = '1 logical core';
        } else if (this.logicalCores === 0 && this.plysicalCores === 1) {
            coresDesc = '1 plysical core';
        } else if (this.logicalCores === 1 && this.plysicalCores === 1) {
            coresDesc = '1 logical core and 1 physical core';
        } else if (this.logicalCores > 1 && this.plysicalCores === 0) {
            coresDesc = `${this.logicalCores} logical cores`;
        } else if (this.logicalCores > 1 && this.plysicalCores === 1) {
            coresDesc = `${this.logicalCores} logical cores and 1 plysical core`;
        } else if (this.logicalCores === 0 && this.plysicalCores > 1) {
            coresDesc = `${this.plysicalCores} plysical cores`;
        } else if (this.logicalCores === 1 && this.plysicalCores > 1) {
            coresDesc = `1 logical core and ${this.plysicalCores} plysical cores`;
        } else if (this.logicalCores > 1 && this.plysicalCores > 1) {
            coresDesc = `${this.logicalCores} logical and ${this.plysicalCores} plysical cores`;
        }

        return [this.name, processorsDesc, coresDesc].filter(Boolean).join(', ');
    }
}
