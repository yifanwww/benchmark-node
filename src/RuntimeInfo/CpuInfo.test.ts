import { CpuInfo } from './CpuInfo';

describe(`Test class ${CpuInfo.name}`, () => {
    it('get data from system', () => {
        expect(() => new CpuInfo()).not.toThrow();

        const cpuInfo = new CpuInfo();
        expect(typeof cpuInfo.name).toBe('string');
        expect(cpuInfo.processors).not.toBeNaN();
        expect(cpuInfo.logicalCores).not.toBeNaN();
        expect(cpuInfo.plysicalCores).not.toBeNaN();
    });

    it('to string', () => {
        function testToString(obj: unknown) {
            expect(CpuInfo.prototype.toString.call(obj)).toMatchSnapshot();
        }

        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 0, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 0, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 0, plysicalCores: 2 });
        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 1, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 1, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 1, plysicalCores: 2 });
        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 2, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 2, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 0, logicalCores: 2, plysicalCores: 2 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 0, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 0, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 0, plysicalCores: 2 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 1, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 1, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 1, plysicalCores: 2 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 2, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 2, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 1, logicalCores: 2, plysicalCores: 2 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 0, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 0, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 0, plysicalCores: 2 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 1, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 1, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 1, plysicalCores: 2 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 2, plysicalCores: 0 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 2, plysicalCores: 1 });
        testToString({ name: '<cpu-name>', processors: 2, logicalCores: 2, plysicalCores: 2 });
    });
});
