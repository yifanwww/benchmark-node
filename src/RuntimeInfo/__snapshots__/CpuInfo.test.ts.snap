// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test class CpuInfo to string 1`] = `"<cpu-name>"`;

exports[`Test class CpuInfo to string 2`] = `"<cpu-name>, 1 plysical core"`;

exports[`Test class CpuInfo to string 3`] = `"<cpu-name>, 2 plysical cores"`;

exports[`Test class CpuInfo to string 4`] = `"<cpu-name>, 1 logical core"`;

exports[`Test class CpuInfo to string 5`] = `"<cpu-name>, 1 logical core and 1 physical core"`;

exports[`Test class CpuInfo to string 6`] = `"<cpu-name>, 1 logical core and 2 plysical cores"`;

exports[`Test class CpuInfo to string 7`] = `"<cpu-name>, 2 logical cores"`;

exports[`Test class CpuInfo to string 8`] = `"<cpu-name>, 2 logical cores and 1 plysical core"`;

exports[`Test class CpuInfo to string 9`] = `"<cpu-name>, 2 logical and 2 plysical cores"`;

exports[`Test class CpuInfo to string 10`] = `"<cpu-name>, 1 CPU"`;

exports[`Test class CpuInfo to string 11`] = `"<cpu-name>, 1 CPU, 1 plysical core"`;

exports[`Test class CpuInfo to string 12`] = `"<cpu-name>, 1 CPU, 2 plysical cores"`;

exports[`Test class CpuInfo to string 13`] = `"<cpu-name>, 1 CPU, 1 logical core"`;

exports[`Test class CpuInfo to string 14`] = `"<cpu-name>, 1 CPU, 1 logical core and 1 physical core"`;

exports[`Test class CpuInfo to string 15`] = `"<cpu-name>, 1 CPU, 1 logical core and 2 plysical cores"`;

exports[`Test class CpuInfo to string 16`] = `"<cpu-name>, 1 CPU, 2 logical cores"`;

exports[`Test class CpuInfo to string 17`] = `"<cpu-name>, 1 CPU, 2 logical cores and 1 plysical core"`;

exports[`Test class CpuInfo to string 18`] = `"<cpu-name>, 1 CPU, 2 logical and 2 plysical cores"`;

exports[`Test class CpuInfo to string 19`] = `"<cpu-name>, 2 CPU"`;

exports[`Test class CpuInfo to string 20`] = `"<cpu-name>, 2 CPU, 1 plysical core"`;

exports[`Test class CpuInfo to string 21`] = `"<cpu-name>, 2 CPU, 2 plysical cores"`;

exports[`Test class CpuInfo to string 22`] = `"<cpu-name>, 2 CPU, 1 logical core"`;

exports[`Test class CpuInfo to string 23`] = `"<cpu-name>, 2 CPU, 1 logical core and 1 physical core"`;

exports[`Test class CpuInfo to string 24`] = `"<cpu-name>, 2 CPU, 1 logical core and 2 plysical cores"`;

exports[`Test class CpuInfo to string 25`] = `"<cpu-name>, 2 CPU, 2 logical cores"`;

exports[`Test class CpuInfo to string 26`] = `"<cpu-name>, 2 CPU, 2 logical cores and 1 plysical core"`;

exports[`Test class CpuInfo to string 27`] = `"<cpu-name>, 2 CPU, 2 logical and 2 plysical cores"`;
